### PR TITLE
MEN-5989: Retrieve signed direct upload URL for artifact

### PR DESCRIPTION
**Pull request by alfrunes**
First part of the direct upload feature done.
I'm deferring the acceptance tests until the whole feature is done.

### CODE REVIEW

**Comment by alfrunes** [comment]
@mender-test-bot start pipeline --fast

**Comment by alfrunes** [comment]
@mender-test-bot start pipeline --fast

File: model/link.go [hunk 1]
```diff
@@ -19,10 +19,11 @@ import (
...
 type Link struct {
-	Uri    string            `json:"uri" bson:"-"`
-	Expire time.Time         `json:"expire,omitempty" bson:"expire"`
-	Method string            `json:"method,omitempty" bson:"-"`
-	Header map[string]string `json:"header,omitempty" bson:"-"`
+	Uri      string            `json:"uri" bson:"-"`
+	Expire   time.Time         `json:"expire,omitempty" bson:"expire"`
+	Method   string            `json:"method,omitempty" bson:"-"`
+	Header   map[string]string `json:"header,omitempty" bson:"-"`
+	TenantID string            `json:"-" bson:"tenant_id"`
```
Comment (kjaskiewiczz) [inline model/link.go:26]: this change shouldn't be part of this commit, right?

File: store/mongo/datastore_mongo.go [hunk 2]
```diff
@@ -838,7 +839,9 @@ func (db *DataStoreMongo) InsertUploadIntent(ctx context.Context, link *model.Up
 	collUploads := db.client.
 		Database(DatabaseName).
 		Collection(CollectionUploadIntents)
-
+	if idty := identity.FromContext(ctx); idty != nil {
```
Comment (kjaskiewiczz) [inline store/mongo/datastore_mongo.go:842]: this change shouldn't be part of this commit, right?

File: model/link.go [hunk 3]
```diff
@@ -19,10 +19,18 @@ import (
 )
 
 type Link struct {
-	Uri    string            `json:"uri"`
-	Expire time.Time         `json:"expire,omitempty"`
-	Method string            `json:"method,omitempty"`
-	Header map[string]string `json:"header,omitempty"`
+	Uri      string            `json:"uri" bson:"-"`
+	Expire   time.Time         `json:"expire,omitempty" bson:"expire"`
+	Method   string            `json:"method,omitempty" bson:"-"`
```
Comment (kjaskiewiczz) [inline model/link.go:24]: where do we use Method and Header?

File: docs/management_api.yml [hunk 4]
```diff
@@ -1183,6 +1183,30 @@ paths:
...
+        - application/json
+      responses:
+        200:
+          description: OK
+          schema:
+            type: object
+            properties:
+              $ref: "#/definitions/ArtifactUploadLink"
+        400:
+          $ref: "#/responses/InvalidRequestError"
```
Comment (merlin-northern) [inline docs/management_api.yml:1204]: idle comment: every time I read it I have the same thought (usually unexpressed): 400 is a _Bad Request_ and should be addressed like that, I checked the desc of this ref, it is not right.

File: docs/management_api.yml [hunk 5]
```diff
@@ -2034,6 +2058,30 @@ definitions:
...
+        description: The ID of the artifact upload intent.
+      uri:
+        type: string
+      expire:
+        type: string
+        format: date-time
+    required:
+      - id
+      - uri
+      - expire
```
Comment (merlin-northern) [inline docs/management_api.yml:2078]: do we restrict somehow the expire? can I set expire to new geological era?

**Review by merlin-northern — APPROVED** [review]
looks good to me, nice. left some comments.

File: docs/management_api.yml (hunk 5 above)
Comment (alfrunes) [reply docs/management_api.yml:2078]: The expiration time is a static configuration, you cannot set it using the API.

File: api/http/api_deployments_test.go [hunk 6]
```diff
@@ -235,6 +235,116 @@ func TestDeploymentsPerTenantHandler(t *testing.T) {
 	}
 }
 
+func TestUploadLink(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply api/http/api_deployments_test.go:238]: We can test it in the acceptance tests, however, it is outside the domain of this service. It is up to S3 or Azure Blob Storage to satisfy the expiration time constraints.

File: model/link.go [hunk 7]
```diff
@@ -19,10 +19,18 @@ import (
...
+	Header   map[string]string `json:"header,omitempty" bson:"-"`
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply model/link.go:25]: Yes, these are ephemeral attributes that we don't persist. We only need to track when the link expires and the object location so we can eventually remove the object.

File: app/app.go [hunk 8]
```diff
@@ -672,6 +673,30 @@ func (d *Deployments) DownloadLink(ctx context.Context, imageID string,
...
+		return nil, errors.WithMessage(err, "app: failed to generate signed URL")
+	}
+	upLink := &model.UploadLink{
+		ArtifactID: artifactID,
+		IssuedAt:   time.Now(),
+		Link:       *link,
+	}
+	err = d.db.InsertUploadIntent(ctx, upLink)
+	if err != nil {
+		return nil, errors.WithMessage(err, "app: error logging the upload intent")
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply app/app.go:694]: IDK: I guess since we're just recording (logging) the URLs metadata (intents) in the database.

File: model/link.go (hunk 3 above)
Comment (alfrunes) [reply model/link.go:24]: It was added as extra metadata to the `storage` client.

**Review by kjaskiewiczz — APPROVED** [review]
looks good
